### PR TITLE
Add CLAMP Functions

### PR DESCRIPTION
**Pull request by lukeqin-s**
Added a macro to clamp a value within specified bounds.

### CODE REVIEW

File: include/common.h [hunk 1]
```diff
@@ -96,4 +96,16 @@ extern void rocketlib_assert_fail_c(const char *file, int line, const char *stat
...
+ * @brief Clamp a value between a lower and upper bound
+ *
+ * @param x The value to clamp
+ * @param lo The lower bound, inclusive
+ * @param hi The upper bound, inclusive
+ *
+ * Note: This macro evaluates its arguments multiple times, so do NOT do clamp(x++, 0, 255)
+ * to avoid incrementing x multiple times.
+ */
+#define CLAMP(x, lo, hi) (((x) < (lo)) ? (lo) : (((x) > (hi)) ? (hi) : (x)))
```
Comment (JasonBrave) [inline include/common.h]: Implement it as a static inline function instead for type safety, name it like `value_clamp_uint32` `value_clamp_float` etc.

File: include/common.h [hunk 2]
```diff
@@ -96,4 +96,16 @@ extern void rocketlib_assert_fail_c(const char *file, int line, const char *stat
...
 #endif
 
+/**
+ * @brief Clamp a value between a lower and upper bound
+ *
+ * @param x The value to clamp
+ * @param lo The lower bound, inclusive
+ * @param hi The upper bound, inclusive
+ *
+ * Note: This macro evaluates its arguments multiple times, so do NOT do clamp(x++, 0, 255)
```
Comment (JasonBrave) [inline include/common.h]: For doc string put description above parameters

**Review by JasonBrave — CHANGES_REQUESTED** [review]
Also you need unit test, create a new file `tests/test_math.cpp`, add thorough test for each inline function, include corner cases. Add the test C++ file to Makefile at root

File: include/common.h (hunk 1 above)
Comment (JasonBrave) [reply include/common.h]: It's a BARR 6.3 violation

File: include/common.h [hunk 3]
```diff
@@ -96,4 +98,40 @@ extern void rocketlib_assert_fail_c(const char *file, int line, const char *stat
...
 
+/**
+ * @brief Clamps a 32 bit unsigned integer between a lower and upper bound.
+ *
+ * @param x The value to clamp
+ * @param lo The lower bound, inclusive
+ * @param hi The upper bound, inclusive
+ * @return The clamped value
+ */
+static inline uint32_t clamp_uint32(uint32_t x, uint32_t lo, uint32_t hi) {
```
Comment (SAT-oO) [inline include/common.h]: can you change the impl style to be consistent with the rest of the file/repo?

namely:
- `float32_t` instead of `float` 
- explicit `if...else...` statements 
- spell out `high` and `low` instead of `hi` and `lo` 

not sure about the `w_` prefix @JasonBrave

**Review by SAT-oO — CHANGES_REQUESTED** [review]

File: include/common.h (hunk 3 above)
Comment (JasonBrave) [reply include/common.h]: > not sure about the w_ prefix

It's for assert only, just use the function names I proposed

> float32_t instead of float

TBH nobody other than canards use floating-point number, so you might just want to keep the float one in canard firmware only
If you really insist on adding floating point number here, you have two options, either use float/double, or adding float32_t/float64_t typedef

File: include/common.h [hunk 4]
```diff
@@ -1,3 +1,5 @@
+#include <stdint.h>
```
Comment (JasonBrave) [inline include/common.h]: Also includes go after include guard

File: include/common.h (hunk 3 above)
Comment (SAT-oO) [reply include/common.h]: we're using the typedef from cmsis in canards repo 
honestly don't care about that in here

File: Makefile [hunk 5]
```diff
@@ -32,7 +32,8 @@ INCLUDE_PATHS := \
 TEST_SRCS := \
 	tests/test_crc8.cpp \
 	tests/test_low_pass_filter.cpp \
-	tests/test_mbr.cpp
+	tests/test_mbr.cpp \
+	tests/test_math.cpp
```
Comment (JasonBrave) [inline Makefile]: Make sure file names are sorted in alphabetical order

**Review by JasonBrave — CHANGES_REQUESTED** [review]

**Review by JasonBrave — CHANGES_REQUESTED** [review]
You need to add float32_t and float64_t 's typedef to common.h, make sure to add static assert to make sure they have corrent width

Then move the clamp functions to a new "mathops.h"

File: rockettest/rockettest.cpp [hunk 6]
```diff
@@ -80,15 +80,15 @@ bool rockettest_test::operator()() {
```
Comment (JasonBrave) [inline rockettest/rockettest.cpp:80]: Do not modify this file, if I did not told you to touch a file then don't touch it

**Review by JasonBrave — CHANGES_REQUESTED** [review]

File: include/mathops.h [hunk 7]
```diff
@@ -0,0 +1,61 @@
+#ifndef MATHOPS_H
+#define MATHOPS_H
+
+#include "common.h"
```
Comment (JasonBrave) [inline include/mathops.h]: order for header files is wrong https://docs.waterloorocketry.com/standards-tutorials/firmware/embedded-coding-standard.html

File: include/mathops.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
+#ifndef MATHOPS_H
```
Comment (JasonBrave) [inline include/mathops.h]:
```suggestion
#ifndef ROCKETLIB_MATHOPS_H
```

File: include/mathops.h [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+#define MATHOPS_H
```
Comment (JasonBrave) [inline include/mathops.h]:
```suggestion
#define ROCKETLIB_MATHOPS_H
```

File: tests/test_math.cpp [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+		rockettest_check_expr_true(value_clamp_int32(7, 7, 7) == 7);
+		rockettest_check_expr_true(value_clamp_int32(100, 7, 7) == 7);
+
+		return test_passed;
+	}
+};
+
+value_clamp_int32_test value_clamp_int32_test_inst;
+
+class value_clamp_float_test : rockettest_test {
```
Comment (JasonBrave) [inline tests/test_math.cpp]:
```suggestion
class value_clamp_float32_test : rockettest_test {
```

File: Makefile [hunk 11]
```diff
@@ -32,6 +33,7 @@ INCLUDE_PATHS := \
 TEST_SRCS := \
 	tests/test_crc8.cpp \
 	tests/test_low_pass_filter.cpp \
+	tests/test_math.cpp \
```
Comment (JasonBrave) [inline Makefile]:
```suggestion
	tests/test_mathops.cpp \
```

Rename the test file to test_mathops.cpp

**Review by JasonBrave — CHANGES_REQUESTED** [review]

**Review by SAT-oO — APPROVED** [review]
lgtm

**Review by JasonBrave — APPROVED** [review]
LGTM. Feel free to merge
